### ISSUE TITLE
`sample` doesn't run in parallel

### ISSUE THREAD

**Issue by teddygroves** [stan-dev/cmdstanpy#103]
#### Summary:
The latest version of cmdstanpy doesn't seem to run the `sample` method in parallel.

#### Description:

To reproduce the issue I made a file in a new folder called `check.py` which is as follows:

```
import cmdstanpy


stan_code = """
data { 
  int<lower=0> N; 
  int<lower=0,upper=1> y[N];
} 
parameters {
  real<lower=0,upper=1> theta;
} 
model {
  theta ~ beta(1,1);
  for (n in 1:N) 
    y[n] ~ bernoulli(theta);
}
"""
data = {
    "N" : 10,
    "y" : [0,1,0,0,0,0,0,0,0,1]
}

with open("bernoulli.stan", 'w') as f:
    f.write(stan_code)

model = cmdstanpy.Model('bernoulli.stan')
model.compile()
fit = model.sample(warmup_iters=100000, cores=2, chains=2, data=data)

```

Then I ran the following shell commands:

```
virtualenv venv
source venv/bin/activate
pip install --no-cache-dir cmdstanpy
install_cmdstan
python check.py
```
Here's the output:

```
~/Code/cmdstanpy_bug $ virtualenv venv
Using base prefix '/Users/tedgro/.pyenv/versions/3.7.2'
New python executable in /Users/tedgro/Code/cmdstanpy_bug/venv/bin/python3.7
Also creating executable in /Users/tedgro/Code/cmdstanpy_bug/venv/bin/python
Installing setuptools, pip, wheel...
done.
~/Code/cmdstanpy_bug $ source venv/bin/activate
(venv) ~/Code/cmdstanpy_bug $ pip install --no-cache-dir cmdstanpy
Collecting cmdstanpy
  Downloading https://files.pythonhosted.org/packages/fa/4c/5fd03a75a71cca63b80b774003e67c8f513d139198f5192b58d4379d1fd7/cmdstanpy-0.4.2-py3-none-any.whl
Collecting pandas (from cmdstanpy)
  Downloading https://files.pythonhosted.org/packages/39/b7/441375a152f3f9929ff8bc2915218ff1a063a59d7137ae0546db616749f9/pandas-0.25.0-cp37-cp37m-macosx_10_9_x86_64.macosx_10_10_x86_64.whl (10.1MB)
     |████████████████████████████████| 10.1MB 731kB/s 
Collecting numpy (from cmdstanpy)
  Downloading https://files.pythonhosted.org/packages/c1/4b/78119133136c20e5ad2e01bf72b0633241defd619939908223cd394a9c32/numpy-1.17.0-cp37-cp37m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl (15.0MB)
     |████████████████████████████████| 15.0MB 776kB/s 
Collecting python-dateutil>=2.6.1 (from pandas->cmdstanpy)
  Downloading https://files.pythonhosted.org/packages/41/17/c62faccbfbd163c7f57f3844689e3a78bae1f403648a6afb1d0866d87fbb/python_dateutil-2.8.0-py2.py3-none-any.whl (226kB)
     |████████████████████████████████| 235kB 3.3MB/s 
Collecting pytz>=2017.2 (from pandas->cmdstanpy)
  Downloading https://files.pythonhosted.org/packages/87/76/46d697698a143e05f77bec5a526bf4e56a0be61d63425b68f4ba553b51f2/pytz-2019.2-py2.py3-none-any.whl (508kB)
     |████████████████████████████████| 512kB 3.1MB/s 
Collecting six>=1.5 (from python-dateutil>=2.6.1->pandas->cmdstanpy)
  Downloading https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
Installing collected packages: six, python-dateutil, pytz, numpy, pandas, cmdstanpy
Successfully installed cmdstanpy-0.4.2 numpy-1.17.0 pandas-0.25.0 python-dateutil-2.8.0 pytz-2019.2 six-1.12.0
(venv) ~/Code/cmdstanpy_bug $ install_cmdstan
CmdStan version: 2.20.0
Install directory: /Users/tedgro/.cmdstanpy
CmdStan version 2.20.0 already installed
deleting tmpfiles dir: /var/folders/nw/nbg0t9fx0hvdhwlgwk2kb48wqkndw3/T/tmp2iqsdetp
done
(venv) ~/Code/cmdstanpy_bug $ python check.py
INFO:cmdstanpy:stan to c++ (bernoulli.hpp)
INFO:cmdstanpy:compiling c++
INFO:cmdstanpy:compiled model file: /Users/tedgro/Code/cmdstanpy_bug/bernoulli
INFO:cmdstanpy:start chain 1
INFO:cmdstanpy:finish chain 1
INFO:cmdstanpy:start chain 2
INFO:cmdstanpy:finish chain 2
deleting tmpfiles dir: /var/folders/nw/nbg0t9fx0hvdhwlgwk2kb48wqkndw3/T/tmpl9f05dxe
done
(venv) ~/Code/cmdstanpy_bug $ 
```
Chain 2 did not start sampling until chain 1 had finished.

#### Current Version:
python: 3.7.2
cmdstan: 2.20.0
cmdstanpy: 0.4.2

**Comment by mitzimorris** [comment]
looking into it now.  thanks for the heads up.

**Comment by ahartikainen** [comment]
Is this a problem also with a bigger model? 

How many cores do you have?

You are running this in macOS?

What compiler do you have?

Sorry for asking so many questions.

**Comment by mitzimorris** [comment]
tested this on my machine using a bigger model and yes, this is a problem.
I'm running MacOs HighSierra - MacBook Pro (Retina, 13-inch, Early 2015) - 4 cores
Python 3.7.0

backing out changes from this commit:
https://github.com/stan-dev/cmdstanpy/commit/d63d4467a9b5d7e9af94c2bd73ec97f2cd95fb22

everything runs in parallel

**Comment by ahartikainen** [comment]
Yes, it happens also on Windows 10.

**Comment by mitzimorris** [comment]
OK, then I'm backing out that change on master.

**Comment by ahartikainen** [comment]
Ok, there is a bug.

    executor.submit(self._do_sample(stanfit, i))

should be really

    executor.submit(self._do_sample, stanfit, i)

After this it is parallel again.

**Comment by ahartikainen** [comment]
it was actually first calling 

    self._do_sample(stanfit, i)

and after it had finished, it send the result to the thread.

**Comment by mitzimorris** [comment]
sorry - missed this message.  reverted to old system.  will fix the fix and add a unit test - same branch, new PR.

**Comment by mitzimorris** [comment]
fixed by https://github.com/stan-dev/cmdstanpy/pull/107
